File: python/src/gwp_py/connection.py
"""GQL connection management."""

from __future__ import annotations

import grpc

from gwp_py.errors import ConnectionError
from gwp_py.session import GqlSession


class GqlConnection:
    """A connection to a GWP server."""

    def __init__(self, channel: grpc.aio.Channel):
        self._channel = channel

    @classmethod
    async def connect(
        cls,
        endpoint: str,
        *,
        credentials: grpc.ChannelCredentials | None = None,
    ) -> GqlConnection:
        """Connect to a GWP server.

        Args:
            endpoint: Server address (e.g. "localhost:50051").
            credentials: Optional TLS credentials.

        Returns:
            A connected GqlConnection.

        Raises:
            ConnectionError: If the connection fails.
        """
        try:
            if credentials is not None:
                channel = grpc.aio.secure_channel(endpoint, credentials)
            else:
                channel = grpc.aio.insecure_channel(endpoint)
            return cls(channel)
        except Exception as e:
            raise ConnectionError(f"failed to connect to {endpoint}: {e}") from e

    async def create_session(self) -> GqlSession:
        """Perform handshake and return a session."""
        return await GqlSession.create(self._channel)

    async def close(self) -> None:
        """Close the underlying gRPC channel."""
        await self._channel.close()

    async def __aenter__(self) -> GqlConnection:
        return self

    async def __aexit__(self, *args: object) -> None:
        await self.close()

File: src/client/connection.rs
//! gRPC connection management.

use tonic::transport::Channel;

use crate::error::GqlError;

use super::GqlSession;

/// A connection to a GQL wire protocol server.
///
/// Manages the gRPC channel and provides session creation.
#[derive(Debug, Clone)]
pub struct GqlConnection {
    channel: Channel,
}

impl GqlConnection {
    /// Connect to a GQL server at the given endpoint.
    ///
    /// # Errors
    ///
    /// Returns an error if the connection cannot be established.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # async fn example() -> Result<(), Box<dyn std::error::Error>> {
    /// use gwp::client::GqlConnection;
    ///
    /// let conn = GqlConnection::connect("http://localhost:50051").await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn connect(endpoint: &str) -> Result<Self, GqlError> {
        let channel = Channel::from_shared(endpoint.to_owned())
            .map_err(|e| GqlError::Protocol(e.to_string()))?
            .connect()
            .await?;

        Ok(Self { channel })
    }

    /// Create a connection from an existing tonic channel.
    #[must_use]
    pub fn from_channel(channel: Channel) -> Self {
        Self { channel }
    }

    /// Perform a handshake and return a session.
    ///
    /// # Errors
    ///
    /// Returns an error if the handshake fails.
    pub async fn create_session(&self) -> Result<GqlSession, GqlError> {
        GqlSession::new(self.channel.clone()).await
    }

    /// Get the underlying tonic channel.
    #[must_use]
    pub fn channel(&self) -> &Channel {
        &self.channel
    }
}
